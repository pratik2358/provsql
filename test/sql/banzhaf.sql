\set ECHO none
SET search_path TO provsql_test,provsql;

CREATE TABLE banzhaf_result AS
  SELECT name, city, banzhaf(c.provenance,p.provenance) FROM (
    SELECT provenance() from (SELECT DISTINCT 1 FROM (
      (SELECT DISTINCT city FROM personnel)
    EXCEPT
      (SELECT p1.city
      FROM personnel p1, personnel p2
      WHERE p1.city = p2.city AND p1.id < p2.id
      GROUP BY p1.city
      ORDER BY p1.city)
      ) t
    ) u)
  AS c,
  (SELECT *, provenance() FROM personnel) AS p;

SELECT remove_provenance('banzhaf_result');

SELECT name, city, ROUND(banzhaf::numeric,3) AS banzhaf FROM banzhaf_result
ORDER BY city, name;

DROP TABLE banzhaf_result;

-- banzhaf computation in the non-probabilistic case
DO $$ BEGIN
  PERFORM set_prob(provenance(), 1.) FROM personnel;
END $$;

CREATE TABLE banzhaf_result AS
  SELECT name, city, banzhaf(c.provenance,p.provenance) FROM (
    SELECT provenance() from (SELECT DISTINCT 1 FROM (
      (SELECT DISTINCT city FROM personnel)
    EXCEPT
      (SELECT p1.city
      FROM personnel p1, personnel p2
      WHERE p1.city = p2.city AND p1.id < p2.id
      GROUP BY p1.city
      ORDER BY p1.city)
      ) t
    ) u)
  AS c,
  (SELECT *, provenance() FROM personnel) AS p;

SELECT remove_provenance('banzhaf_result');

SELECT name, city, ROUND(banzhaf::numeric,3) AS banzhaf FROM banzhaf_result
ORDER BY city, name;

DROP TABLE banzhaf_result;

-- Put back original probability values
DO $$ BEGIN
  PERFORM set_prob(provenance(), id*1./10) FROM personnel;
END $$;

CREATE TABLE banzhaf_result1 AS
    SELECT city, provenance() FROM (
       (SELECT DISTINCT city FROM personnel)
     EXCEPT
       (SELECT p1.city
       FROM personnel p1, personnel p2
       WHERE p1.city = p2.city AND p1.id < p2.id
       GROUP BY p1.city
       ORDER BY p1.city)
       ) t;
SELECT remove_provenance('banzhaf_result1');
CREATE TABLE banzhaf_result2 AS
  SELECT * FROM banzhaf_result1, banzhaf_all_vars(provenance);

SELECT city, ROUND(value::numeric,3) AS banzhaf FROM banzhaf_result2
ORDER BY city, banzhaf;

DROP TABLE banzhaf_result1;
DROP TABLE banzhaf_result2;
