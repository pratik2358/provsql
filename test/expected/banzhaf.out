\set ECHO none
 remove_provenance 
-------------------
 
(1 row)

   name   |   city   | banzhaf 
----------+----------+---------
 Ellen    | Berlin   |   0.211
 Susan    | Berlin   |   0.737
 John     | New York |   0.176
 Paul     | New York |   0.396
 Dave     | Paris    |  -0.039
 Magdalen | Paris    |   0.065
 Nancy    | Paris    |   0.157
(7 rows)

 remove_provenance 
-------------------
 
(1 row)

   name   |   city   | banzhaf 
----------+----------+---------
 Ellen    | Berlin   |   0.000
 Susan    | Berlin   |   0.000
 John     | New York |   0.000
 Paul     | New York |   0.000
 Dave     | Paris    |  -4.000
 Magdalen | Paris    |  -4.000
 Nancy    | Paris    |  -4.000
(7 rows)

 remove_provenance 
-------------------
 
(1 row)

   city   | banzhaf 
----------+---------
 Berlin   |   0.120
 Berlin   |   0.420
 New York |   0.080
 New York |   0.180
 Paris    |  -0.030
 Paris    |   0.050
 Paris    |   0.120
(7 rows)

